Lab-4/HelmerNylen_lab4.ipynb

```diff
@@ -0,0 +1,422 @@
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "A9A-Sr0dHHgO",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "source": [
+        "# Represents a continuous pointwise linear approximation\n",
+        "class Approx:\n",
+        "  def __init__(self, alpha, mesh):\n",
+        "    self.alpha = alpha\n",
+        "    self.mesh = mesh\n",
+        "  \n",
+        "  def __call__(self, x):\n",
+        "    assert self.mesh[0] <= x <= self.mesh[-1]\n",
+        "    if x == self.mesh[-1]:\n",
+        "      return self.alpha[-1]\n",
+        "    \n",
+        "    i = (self.mesh <= x).nonzero()[0][-1]\n",
+        "    t = (x - self.mesh[i]) / (self.mesh[i+1] - self.mesh[i])\n",
+        "    return self.alpha[i] + (self.alpha[i+1] - self.alpha[i]) * t"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "N-fAUVwBZJYO",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "source": [
+        "def _repeated(xs, reps):\n",
+        "  for x, rep in zip(xs, reps):\n",
+        "    for _ in range(rep):\n",
+        "      yield x\n",
+        "\n",
+        "def test_1D():\n",
+        "  # Produce a random polynomial on [-10, 10] and use as f\n",
+        "  n_zeros = randint(2, 4)\n",
+        "  zeros = np.random.rand(n_zeros) * 20 - 10\n",
+        "  multiplicities = [randint(1, 2) for i in range(n_zeros)]\n",
+        "  polynomials = [(-zeros[i], 1) for i in range(n_zeros)]\n",
+        "  pol = reduce(P.polymul, _repeated(polynomials, multiplicities))\n",
+        "  pol = pol / np.sqrt(max(pol))\n",
+        "  f = lambda x: P.polyval(x, pol)\n",
+        "\n",
+        "  errors = []\n",
+        "  fine_mesh = np.linspace(-10, 10, 1500)\n",
+        "  n_meshes = range(10, 1100, 100)\n",
+        "\n",
+        "  # Test a number of different meshes\n",
+        "  for n_mesh in n_meshes:\n",
+        "    mesh = np.unique(np.concatenate(([-10, 10], np.random.rand(n_mesh) * 20 - 10)))\n",
+        "\n",
+        "    alpha = L2_proj_1D(f, mesh)\n",
+        "    f_approx = Approx(alpha, mesh)\n",
+        "    # Assert that the approximation is valid for each mesh point (accuracy)\n",
+        "    assert np.allclose(0, alpha - [f(x) for x in mesh])\n",
+        "    \n",
+        "    errors.append(sum(abs(f(x) - f_approx(x)) for x in fine_mesh))\n",
+        "\n",
+        "  # Assert that the error decreases with a finer mesh (convergence)\n",
+        "  k, m = np.polyfit(n_meshes, np.log(errors), 1)\n",
+        "  assert k < 0"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "SsQLT38gVbn_",
+        "colab_type": "text"
+      },
+      "source": [
+        "# **Results**"
```

Comment:
You need to add a plot showing a comparison between your approximation and the exact solution, as well as discuss/show the convergence of the approximation.

Reply:
Done